File: ios/FaceLivenessDetectorCredentials.swift
import AWSPluginsCore
import Foundation

enum FaceLivenessDetectorCredentialStoreError: LocalizedError {
  case missingCredentials
  case expiredCredentials
  case invalidExpiration

  var errorDescription: String? {
    switch self {
    case .missingCredentials:
      return "Missing Face Liveness credentials"
    case .expiredCredentials:
      return "Face Liveness credentials are expired or close to expiry"
    case .invalidExpiration:
      return "Face Liveness credential expiration must be a valid ISO-8601 date"
    }
  }
}

struct FaceLivenessDetectorAWSTemporaryCredentials: AWSTemporaryCredentials {
  let accessKeyId: String
  let secretAccessKey: String
  let sessionToken: String
  let expiration: Date

  var isCloseToExpiry: Bool {
    expiration.timeIntervalSinceNow <= 60
  }
}

final class FaceLivenessDetectorCredentialStore {
  private static let lock = NSLock()
  private static var credentials: FaceLivenessDetectorAWSTemporaryCredentials?

  static func set(_ newCredentials: FaceLivenessDetectorAWSTemporaryCredentials) throws {
    lock.lock()
    defer { lock.unlock() }

    guard !newCredentials.accessKeyId.isEmpty,
      !newCredentials.secretAccessKey.isEmpty,
      !newCredentials.sessionToken.isEmpty
    else {
      throw FaceLivenessDetectorCredentialStoreError.missingCredentials
    }

    guard !newCredentials.isCloseToExpiry else {
      throw FaceLivenessDetectorCredentialStoreError.expiredCredentials
    }

    credentials = newCredentials
  }

  static func clear() {
    lock.lock()
    defer { lock.unlock() }

    credentials = nil
  }

  static func getCredentials() throws -> FaceLivenessDetectorAWSTemporaryCredentials {
    lock.lock()
    defer { lock.unlock() }

    guard let current = credentials else {
      throw FaceLivenessDetectorCredentialStoreError.missingCredentials
    }

    guard !current.isCloseToExpiry else {
      throw FaceLivenessDetectorCredentialStoreError.expiredCredentials
    }

    return current
  }
}

final class FaceLivenessDetectorCredentialsProvider: AWSCredentialsProvider {
  func fetchAWSCredentials() async throws -> AWSCredentials {
    try FaceLivenessDetectorCredentialStore.getCredentials()
  }
}

func parseFaceLivenessDetectorExpiration(_ value: String) throws -> Date {
  let isoFormatter = ISO8601DateFormatter()
  isoFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

  if let date = isoFormatter.date(from: value) {
    return date
  }

  isoFormatter.formatOptions = [.withInternetDateTime]

  if let date = isoFormatter.date(from: value) {
    return date
  }

  throw FaceLivenessDetectorCredentialStoreError.invalidExpiration
}

File: android/src/main/java/expo/modules/faceliveness/FaceLivenessDetectorCredentials.kt
package expo.modules.faceliveness

import aws.smithy.kotlin.runtime.time.Instant as SmithyInstant
import com.amplifyframework.auth.AWSCredentials
import com.amplifyframework.auth.AWSCredentialsProvider
import com.amplifyframework.auth.AWSTemporaryCredentials
import com.amplifyframework.auth.AuthException
import com.amplifyframework.core.Consumer
import java.time.Instant

data class FaceLivenessDetectorAWSTemporaryCredentials(
  val accessKeyId: String,
  val secretAccessKey: String,
  val sessionToken: String,
  val expiration: String
) {
  val expirationInstant: Instant
    get() = Instant.parse(expiration)

  fun isCloseToExpiry(): Boolean {
    return expirationInstant.minusSeconds(60).isBefore(Instant.now())
  }
}

object FaceLivenessDetectorCredentialStore {
  private val lock = Any()
  private var credentials: FaceLivenessDetectorAWSTemporaryCredentials? = null

  fun set(newCredentials: FaceLivenessDetectorAWSTemporaryCredentials) {
    synchronized(lock) {
      if (
        newCredentials.accessKeyId.isBlank() ||
        newCredentials.secretAccessKey.isBlank() ||
        newCredentials.sessionToken.isBlank()
      ) {
        throw IllegalStateException("Missing Face Liveness credentials")
      }

      if (newCredentials.isCloseToExpiry()) {
        throw IllegalStateException("Face Liveness credentials are expired or close to expiry")
      }

      credentials = newCredentials
    }
  }

  fun clear() {
    synchronized(lock) {
      credentials = null
    }
  }

  fun getCredentials(): FaceLivenessDetectorAWSTemporaryCredentials {
    synchronized(lock) {
      val current = credentials
        ?: throw IllegalStateException("Missing Face Liveness credentials")

      if (current.isCloseToExpiry()) {
        throw IllegalStateException("Face Liveness credentials are expired or close to expiry")
      }

      return current
    }
  }
}

class FaceLivenessDetectorCredentialsProvider : AWSCredentialsProvider<AWSCredentials> {
  override fun fetchAWSCredentials(
    onSuccess: Consumer<AWSCredentials>,
    onError: Consumer<AuthException>
  ) {
    try {
      val credentials = FaceLivenessDetectorCredentialStore.getCredentials()

      onSuccess.accept(
        AWSTemporaryCredentials(
          accessKeyId = credentials.accessKeyId,
          secretAccessKey = credentials.secretAccessKey,
          sessionToken = credentials.sessionToken,
          expiration = SmithyInstant(credentials.expirationInstant)
        )
      )
    } catch (error: Exception) {
      onError.accept(
        AuthException(
          "Not authorized to perform a face liveness check.",
          "Valid credentials are required for the face liveness check.",
          error
        )
      )
    }
  }
}
